test/test_log_forwarding.py

```diff
@@ -0,0 +1,150 @@
+import gzip
+from io import (
+    StringIO,
+)
+import json
+from unittest.mock import (
+    MagicMock,
+    patch,
+)
+
+import chalice.app
+from moto import (
+    mock_s3,
+)
+
+from app_test_case import (
+    LambdaEnvTestCase,
+)
+from azul import (
+    cached_property,
+    config,
+)
+from azul.deployment import (
+    aws,
+)
+from azul.indexer.log_forwarding_controller import (
+    LogForwardingController,
+)
+from azul.service.storage_service import (
+    StorageService,
+)
+
+
+class TestLogForwarding(LambdaEnvTestCase):
+
+    @property
+    def log_bucket(self) -> str:
+        return aws.qualified_bucket_name(config.logs_term)
+
+    @property
+    def log_file_key(self) -> str:
+        prefix = config.alb_access_log_path_prefix('gitlab')
+        return '/'.join([prefix, 'AWSLogs', '123123123123', 'elasticloadbalancing', '2023', '01', '01', 'test.log.gz'])
+
+    @cached_property
+    def storage_service(self) -> StorageService:
+        return StorageService(bucket_name=self.log_bucket)
+
+    @mock_s3
+    def test(self):
+        self.storage_service.create_bucket()
+        event = chalice.app.S3Event(context={}, event_dict={
+            'Records': [{
+                's3': {
+                    'bucket': {'name': self.log_bucket},
+                    'object': {'key': self.log_file_key}
+                }
+            }]
+        })
+
+        escaped_chars = {
+            '"': r'\x22',
+            '\\': r'\x5c',
+            '\U0001F600': r'\xf0\x9f\x98\x80',
```

Comment:
Python source code is assumed to be encoded in UTF-8 by default and you don't need to resort to unicode escapes on the left-hand side. The right hand-side is a little obscure and needs an explanation. It looks like the four bytes of the resulting UTF-8 encoding, then escaped using the proprietary `\xXX` syntax.

You probably want to use a unicode letter like `ü`, not an emoticon. Letters have smaller code points, and produce shorter UTF-8.

Reply:
The RHS is how the LHS is encoded in the log file. In other words, if the LHS is what we put into the request, the RHS is what the ALB logger spits out.